bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringResultScanner.java

```diff
@@ -0,0 +1,238 @@
+                @Override
+                public Result call() throws IOException {
+                  return MirroringResultScanner.this.primaryResultScanner.next();
+                }
+              },
+              HBaseOperation.NEXT);
+
+      int startingIndex = this.readEntries;
+      this.readEntries += 1;
+      ScannerRequestContext context =
+          new ScannerRequestContext(
+              this.originalScan,
+              result,
+              startingIndex,
+              this.mirroringTracer.spanFactory.getCurrentSpan());
+
+      this.scheduleRequest(
+          new RequestResourcesDescription(result),
+          this.secondaryResultScannerWrapper.next(context),
+          this.verificationContinuationFactory.scannerNext());
+      return result;
+    }
+  }
+
+  @Override
+  public Result[] next(final int entriesToRead) throws IOException {
+    try (Scope scope =
+        this.mirroringTracer.spanFactory.operationScope(HBaseOperation.NEXT_MULTIPLE)) {
+      Result[] results =
+          this.mirroringTracer.spanFactory.wrapPrimaryOperation(
+              new CallableThrowingIOException<Result[]>() {
+                @Override
+                public Result[] call() throws IOException {
+                  return MirroringResultScanner.this.primaryResultScanner.next(entriesToRead);
+                }
+              },
+              HBaseOperation.NEXT_MULTIPLE);
+
+      // TODO: remove this index, it doesn't tell the user anything.
+      int startingIndex = this.readEntries;
+      this.readEntries += entriesToRead;
+      ScannerRequestContext context =
+          new ScannerRequestContext(
+              this.originalScan,
+              results,
+              startingIndex,
+              entriesToRead,
+              this.mirroringTracer.spanFactory.getCurrentSpan());
+      this.scheduleRequest(
+          new RequestResourcesDescription(results),
+          this.secondaryResultScannerWrapper.next(context),
+          this.verificationContinuationFactory.scannerNext());
+      return results;
+    }
+  }
+
+  @Override
+  public void close() {
+    if (this.closed.getAndSet(true)) {
+      return;
+    }
+
+    AccumulatedExceptions exceptionsList = new AccumulatedExceptions();
+    try {
+      this.primaryResultScanner.close();
+    } catch (RuntimeException e) {
+      exceptionsList.add(e);
+    } finally {
+      try {
+        this.asyncClose();
+      } catch (RuntimeException e) {
+        log.error("Exception while scheduling this.close().", e);
+        exceptionsList.add(e);
+      }
+    }
+
+    exceptionsList.rethrowAsRuntimeExceptionIfCaptured();
+  }
+
+  @VisibleForTesting
+  ListenableFuture<Void> asyncClose() {
+    this.secondaryResultScannerWrapper.asyncClose();
+    this.listenableReferenceCounter.decrementReferenceCount();
+    return this.listenableReferenceCounter.getOnLastReferenceClosed();
+  }
+
+  @Override
+  public boolean renewLease() {
```

Comment:
As discussed elsewhere, leases are not needed and not supported by bigtable client (renewLease throws UnsupportedOperationException but it might as well `return true` instead of doing that). Improved renewLease ignores UnsupportedOperationException. Honestly, I'm not a HBase expert myself and I have no idea how often code in the wild uses this method. I assume that if we renew the lease it will expire automatically after some time and we do not have to clean it up if only one of the leases was renewed.